tasks/track_memory_allocations/src/lib.rs

```diff
@@ -73,16 +74,108 @@ fn test() {
     run().unwrap();
 }
 
+struct Stats<'a> {
+    file: &'a TestFile,
+    sys_allocs: usize,
+    sys_reallocs: usize,
+    arena_allocs: usize,
+    arena_reallocs: usize,
+    arena_bytes: usize,
+}
+
+fn record_stats<'a>(file: &'a TestFile, allocator: &Allocator) -> Stats<'a> {
+    let sys_allocs = NUM_ALLOC.load(SeqCst);
+    let sys_reallocs = NUM_REALLOC.load(SeqCst);
+    #[cfg(not(feature = "is_all_features"))]
+    let (arena_allocs, arena_reallocs) = allocator.get_allocation_stats();
+    #[cfg(feature = "is_all_features")]
+    let (arena_allocs, arena_reallocs) = (0, 0);
+    let arena_bytes = allocator.used_bytes();
+
+    Stats { file, sys_allocs, sys_reallocs, arena_allocs, arena_reallocs, arena_bytes }
+}
+
 /// # Panics
 /// # Errors
 pub fn run() -> Result<(), io::Error> {
     let files = TestFiles::complicated();
-    let snap_path = project_root().join("tasks/track_memory_allocations/allocs_parser.snap");
 
+    let mut allocator = Allocator::default();
+    let scratch_allocator = Allocator::default();
+
+    let options = ParseOptions { parse_regular_expression: true, ..ParseOptions::default() };
+
+    let mut parser_stats = Vec::new();
+    let mut semantic_stats = Vec::new();
+
+    // Warm-up by parsing each file first, and then measuring the actual allocations. This reduces variance
+    // in the number of allocations, because we ensure that the bump allocator has already requested all
+    // of the space it will need from the system allocator to parse the largest file in the set.
+    for file in files.files() {
+        let ret = Parser::new(&allocator, &file.source_text, file.source_type)
+            .with_options(options)
+            .parse();
+        SemanticBuilder::new()
+            .with_check_syntax_error(false)
+            .with_scope_tree_child_ids(true)
+            .with_cfg(true)
+            .with_build_jsdoc(true)
+            .build(&ret.program);
+    }
+
+    for file in files.files() {
+        allocator.reset();
+        reset_global_allocs();
+
+        let ret = Parser::new(&allocator, &file.source_text, file.source_type)
+            .with_options(options)
+            .parse();
+
+        parser_stats.push(record_stats(file, &allocator));
+
+        let program = ret.program.clone_in(&scratch_allocator);
```

Comment:
Can we skip cloning `Program` here?

The cloned `Program` will be slightly different from the original - e.g. a `Vec` with len 1, cap 4 will be compacted in the clone to len 1, cap 1. Working with that compacted AST *might* alter the pattern of heap allocations slightly. By my logic, it *shouldn't* make any difference, but there's other things we're seeing with heap allocations that I can't explain (different behavior on different platforms), so obviously my logic is wrong somehow. So I think its preferable to stick as close to "real world" usage as we can.

A more obvious reason to remove the clone is that I don't think it does anything useful, since we're not able to measure arena allocations in semantic at present. And if we could, it'd be a different allocator anyway.

Reply:
Also, I don't think you're `reset`-ing the `scratch_allocator`.

Reply:
This was some laziness on my part, I was running into borrow checking issues and I tried to clone my way into some productivity. Looking at this with fresh eyes, I realized it's because we were resetting the allocator between parser/semantic which will obviously clear memory for the program, so we don't want to do that. I updated so that we can compute the stats by taking diffs from the previous allocation state and only doing a reset at the beginning of each loop.

Reply:
Ah the dastardly borrow-checker, always peeing on our parade.

Building `Semantic` doesn't make any allocations into to AST `Allocator`, does it? Does it??? If it does, you've discovered something we should address.